src/lib_ccx/ccx_encoders_helpers.c

```diff
@@ -357,84 +428,74 @@ used=encode_line (ctx->buffer,(unsigned char *) string);
 fwrite (ctx->buffer,used,1,fh);
 }*/
 
-int add_word(const char *word)
+// Add the word the `struct word_list` list
+int add_word(struct word_list *list, const char *word)
 {
-	char *new_lower;
-	char *new_correct;
-	char **ptr_lower;
-	char **ptr_correct;
-	int i;
-	if (spell_words == spell_capacity)
+	if (list->len == list->capacity)
 	{
-		// Time to grow
-		spell_capacity += 50;
-		ptr_lower = (char **)realloc(spell_lower, sizeof (char *)*
-			spell_capacity);
-		ptr_correct = (char **)realloc(spell_correct, sizeof (char *)*
-			spell_capacity);
-	}
-	else
-	{
-		ptr_lower = spell_lower;
-		ptr_correct = spell_correct;
-	}
-	size_t len = strlen(word);
-	new_lower = (char *)malloc(len + 1);
-	new_correct = (char *)malloc(len + 1);
-	if (ptr_lower == NULL || ptr_correct == NULL ||
-		new_lower == NULL || new_correct == NULL)
-	{
-		spell_capacity = 0;
-		for (i = 0; i < spell_words; i++)
+		list->capacity += 50;
+		if ((list->words = realloc(list->words, list->capacity * sizeof(char *))) == NULL)
 		{
-			freep(&spell_lower[spell_words]);
-			freep(&spell_correct[spell_words]);
+			return -1;
 		}
-		freep(&spell_lower);
-		freep(&spell_correct);
-		freep(&ptr_lower);
-		freep(&ptr_correct);
-		freep(&new_lower);
-		freep(&new_correct);
-		spell_words = 0;
-		return -1;
 	}
-	else
+
+	size_t word_len = strlen(word);
+	if ((list->words[list->len] = malloc(word_len + 1)) == NULL)
 	{
-		spell_lower = ptr_lower;
-		spell_correct = ptr_correct;
+		return -1;
 	}
-	strcpy(new_correct, word);
-	for (size_t i = 0; i<len; i++)
+
+	strcpy(list->words[list->len++], word);
+	return word_len;
+}
+
+int add_builtin_capitalized_words()
+{
+	static int function_already_ran = 0; // so we don't do it twice
```

Comment:
You just made the function multithread unsafe :-)
Do you really need this? Seems like not calling it twice is enough...

Reply:
I just translated what was previously there.

Reply:
The problem is that someone could pass the `--kf` or `-sc` argument twice.

How could I make it multithread safe?

Reply:
In general static variables in a function make it MT unsafe until there's a mutex involved (and we definitely don't want that extra complexity). Anyway I think I'm OK-ish with that here.

Reply:
OK, it was already like that before.

Reply:
Actually I wasn't being careful, these functions cannot be called twice as they are dealt with after the parsing of arguments.